File: codigo/matlab/gauss_jordan.m
function [x, Inv] = gauss_jordan(A, b)
% GAUSS_JORDAN  Gauss-Jordan con pivoteo: solucion e inversa. Chapra cap. 9-10.
%   Uso:  [x, Inv] = gauss_jordan([3 -0.1 -0.2; 0.1 7 -0.3; 0.3 -0.2 10], ...
%                                 [7.85; -19.3; 71.4]);
    n = size(A,1);
    M = [A, eye(n), b(:)];                    % [A | I | b]
    for k = 1:n
        [~, p] = max(abs(M(k:n, k))); p = p + k - 1;
        if p ~= k, M([k p], :) = M([p k], :); end
        M(k, :) = M(k, :)/M(k, k);            % normaliza el pivote
        for i = 1:n
            if i ~= k
                M(i, :) = M(i, :) - M(i, k)*M(k, :);
            end
        end
    end
    Inv = M(:, n+1:2*n);
    x   = M(:, end);
end

File: codigo/python/gauss_jordan.py
"""gauss_jordan.py -- Gauss-Jordan con pivoteo: solucion e inversa. Chapra cap. 9-10."""


def gauss_jordan(A, b):
    n = len(b)
    # matriz aumentada [A | I | b]
    M = [A[i][:] + [1.0 if i == j else 0.0 for j in range(n)] + [b[i]]
         for i in range(n)]
    for k in range(n):
        p = max(range(k, n), key=lambda i: abs(M[i][k]))   # pivoteo parcial
        if p != k:
            M[k], M[p] = M[p], M[k]
        piv = M[k][k]
        M[k] = [v / piv for v in M[k]]                     # normaliza el pivote
        for i in range(n):
            if i != k:
                factor = M[i][k]
                M[i] = [M[i][j] - factor * M[k][j] for j in range(len(M[i]))]
    inv = [row[n:2 * n] for row in M]
    x = [row[-1] for row in M]
    return x, inv


if __name__ == "__main__":
    A = [[3, -0.1, -0.2], [0.1, 7, -0.3], [0.3, -0.2, 10]]
    b = [7.85, -19.3, 71.4]
    x, inv = gauss_jordan(A, b)
    print("Solucion:", [round(v, 6) for v in x])
